ls1intum/Artemis
# `Lectures`: Fix the order of lecture units when merging pdf files

### Checklist
#### General
#### Server

### Motivation and Context
Closes #6645

### Description
In #4972 I implemented that in the merged pdf of all attachment units the pages should follow the order of the units.

Later #5099 fixed a race-condition by adding `@Transient` to the attribute, which broke this feature. Now, hibernate ignores the saved value when building the object, its value is always 0.

To combine both PRs I now completely removed the individual order in the attachment unit. It still gets saved in the database and the list used in the lecture follows this order. When merging the PDF file, we now retrieve the a ordered list directly from the database.

I was not able to retrieve the order without using the lecture in the query. This is the most performance-optimal query I was able to write here, if you have any other ideas on how this could be improved please let me know. You can find more information on the used INDEX and TREAT functions here: https://en.wikibooks.org/wiki/Java_Persistence/JPQL#Special_Operators


### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Student
- 1 Lecture with multiple .pdf attachment units

1. Student: Download the merged pdf file for the lecture
2. Instructor: Re-order the units.
3. Student: Download the merged pdf again, verify the changed order
4. Repeat that process a few times and check that race conditions as in #5099 are not noticable


### Review Progress
#### Performance Review
#### Code Review
#### Manual Tests
